CrewChiefV4/RF2/RF2Spotter.cs

```diff
@@ -55,6 +55,16 @@ private rF2VehicleScoring getVehicleInfo(CrewChiefV4.rFactor2.RF2SharedMemoryRea
             throw new Exception("no vehicle for player!");
         }
 
+        private bool hasVehicleInfo(CrewChiefV4.rFactor2.RF2SharedMemoryReader.RF2StructWrapper shared)
+        {
+            for (int i = 0; i < shared.scoring.mScoringInfo.mNumVehicles; ++i)
```

Comment:
Thanks for this.   I don't think we can do this function on every spotter call.  The better fix would be to set some flag in RF2 game state mapper to indicate if we have player info or not, because it does not execute this often.   Also, I remember there was a reason why FC version of realtime was used, and not the scoring flag, but I don't recall exact reason, so this might have to be revisited.

Did you verify this fixes the issue?

Reply:
I tested that this fixes the lack of car left/car right/clear/etc. messages after a driver swap and also that it still works offline.

Reply:
I just had a quick look at a code.  I have nothing against this change, but it isn't complete.  I see several spots in mapper that insist on both inRealTime values being true or false at the same time.  rF2 mapper is pretty polished, so I doubt that was done by mistake.  Sure, this fix will fix spotter case.  But in order to find a complete solution we need a trace to understand what exactly happen to both values, and adjust rF2 logic everywhere, not just in the spotter.

Reply:
Going through the RF2 mapper, I see 3 places where we rely on extended.mInRealtimeFC and scoring.mScoringInfo.mInRealtime both being in sync. 2 of these are fairly minor - pit crew 'done' message and engine stalled message will only trigger if both these values are 1. 

The one that's more concerning is where the app relies on these both being 1 in order to detect a n abrupt session restart - specifically when you restart a session without going back to the monitor first. I think that, after you've done your driver swaps, engine stall and pit crew ready will no longer play and hitting 'esc' and selecting 'restart session' will make the app think the session is completed. Not sure why you'd ever do this tho :).

So it looks (to me at least) like only checking 1 of the realTime values in the spotter *might* be OK. I guess the biggest risk here is that the spotter runs when it really shouldn't be running (in replays or spectator mode, for example).

I agree that the hasVehicleInfo logic should be in the mapper simply because we're already iterating this array in the mapper and it's less overhead. Also, do we actually need to check this value for the previous state as well as the current state? Maybe we can just add a public static bool in the mapper that's set on each tick for the current game state and check that in the spotter.

Also, thanks for contributing and welcome to the CC team :)

Reply:
just to illustrate what we mean about doing the 'has player data' check in the mapper, something like this:

https://github.com/mrbelowski/CrewChiefV4/tree/rf2-spotter-driver-swap

Reply:
The array is already being iterated by the spotter a bit further down so I adjusted it so it's not done twice. 
The previous state does need to be checked as the spotter is using some values from it. Your linked example is kind of checking that by catching an exception from getVehicleInfo. I've adjusted it so there's no need for the exception.

Reply:
this is a lot better. I'm happy to merge it but we need 1 change - probably a c# older syntax thing (I build with VS 2013) - the currentPlayerScoring and previousPlayerScoring variables need to be declared before the if block:

```
            rF2VehicleScoring currentPlayerScoring;
            rF2VehicleScoring previousPlayerScoring;

            if (!this.enabled 
                || currentState.scoring.mScoringInfo.mCurrentET < this.timeAfterRaceStartToActivate
                || currentGameState.OpponentData.Count == 0
                || currentState.scoring.mScoringInfo.mInRealtime == 0
                || lastState.scoring.mScoringInfo.mInRealtime == 0
                // turn off spotter for formation lap before going green
                || currentState.scoring.mScoringInfo.mGamePhase == (int)rFactor2Constants.rF2GamePhase.Formation
                || !tryGetVehicleInfo(currentState, out currentPlayerScoring)
                || !tryGetVehicleInfo(lastState, out previousPlayerScoring))
                return;
```